acceptance/lib/puppet/acceptance/common_utils.rb

```diff
@@ -6,5 +6,108 @@ def ruby_command(host)
       end
       module_function :ruby_command
     end
-  end
-end
+
+    module CAUtils
+      def initialize_ssl
+        hostname = on(master, 'facter hostname').stdout.strip
+        fqdn = on(master, 'facter fqdn').stdout.strip
+
+        step "Ensure puppet is stopped"
+        # must be shutdown for the cert setup steps to work, running puppet master will
+        # potentially ignore the puppet.conf changes.
+        on(master, puppet('resource', 'service', 'puppetserver', "ensure=stopped"))
+
+        step "Clear SSL on all hosts"
+        hosts.each do |host|
+          ssldir = on(host, puppet('agent --configprint ssldir')).stdout.chomp
+          on(host, "rm -rf '#{ssldir}'")
+        end
+
+        step "Master: Start Puppet Master" do
+          master_opts = {
+            :main => {
+              :dns_alt_names => "puppet,#{hostname},#{fqdn}",
+            },
+            :__service_args__ => {
+              # apache2 service scripts can't restart if we've removed the ssl dir
+              :bypass_service_script => true,
+            },
+          }
+          with_puppet_running_on(master, master_opts) do
+
+            hosts.each do |host|
+              next if host['roles'].include? 'master'
+
+              step "Agents: Run agent --test first time to gen CSR"
+              on host, puppet("agent --test --server #{master}"), :acceptable_exit_codes => [1]
+            end
+
+            # Sign all waiting certs
+            step "Master: sign all certs"
+            on master, puppet("cert --sign --all"), :acceptable_exit_codes => [0,24]
+
+            step "Agents: Run agent --test second time to obtain signed cert"
+            on agents, puppet("agent --test --server #{master}"), :acceptable_exit_codes => [0,2]
```

Comment:
I think if you move `050_StartPuppetServer.rb` to be `035_StartPuppetServer.rb` so that it comes before `040_ValidateSignCert.rb`, then the puppetserver will generate a cert/key of its own as part of its normal installation process. Then you shouldn't need to wipe out the SSL directory on the master host and start webrick (via `with_puppet_running_on`) to regenerate the master's SSL cert.

So then this method can be a shorter version of what you have:

```ruby
agents.each do |agent|
  next if agent == master

  ssldir = on(agent, puppet('agent --configprint ssldir')).stdout.chomp
  on(agent, "rm -rf '#{ssldir}'")
  
  step "Agents: Run agent --test first time to gen CSR"
  on agent, puppet("agent --test --server #{master}"), :acceptable_exit_codes => [1]
end

step "Master: sign all certs"
on master, puppet("cert --sign --all"), :acceptable_exit_codes => [0,24]

agents.each do |agent|
  next if agent == master

  step "Agents: Run agent --test second time to obtain signed cert"
  on agent, puppet("agent --test --server #{master}"), :acceptable_exit_codes => [0,2]
end
```